File: R-2.10.1/interpreter/topic-types.R

source( 'testEnv.R')

te = TestEnv()

te$checkComplainAndAdjustExpected( 0)

te$test( "logical"==mode( FALSE))
te$test( "logical"==mode( TRUE))
te$test( "logical"==typeof( TRUE))
te$test( "logical"==class( TRUE))
te$test( "NULL"==mode( NULL))
te$test( "NULL"==typeof( NULL))
te$test( "NULL"==class( NULL))
te$test( "numeric"==mode( 1))
te$test( "double"==typeof( 1))
te$test( "numeric"==class( 1))
te$test( "numeric"==mode( 1.5))
te$test( "double"==typeof( 1.5))
te$test( "numeric"==class( 1.5))
te$test( "numeric"==mode( 1L))
te$test( "integer"==typeof( 1L))
te$test( "integer"==class( 1L))
te$test( "character"==mode( "ab"))
te$test( "character"==typeof( "ab"))
te$test( "character"==class( "ab"))
te$test( "character"==mode( 'ab'))
te$test( "character"==typeof( 'ab'))
te$test( "character"==class( 'ab'))

te$test( "NULL"==mode( c()))
te$test( "NULL"==mode( c( NULL, NULL)))
te$test( "NULL"==typeof( c( NULL, NULL)))
te$test( "logical"==mode( c( TRUE, FALSE)))
te$test( "logical"==mode( c( TRUE, FALSE, NULL)))
te$test( "logical"==typeof( c( TRUE, FALSE, NULL)))
te$test( "numeric"==mode( c( 2, 3, 4)))
te$test( "numeric"==mode( c( 2, 3, 4, TRUE, NULL)))
te$test( "double"==typeof( c( 2, 3, 4, TRUE, NULL)))
te$test( "double"==typeof( c( 2L, 3L, 4, TRUE, NULL)))
te$test( "integer"==typeof( c( 2L, 3L, 4L, TRUE, NULL)))
te$test( "character"==mode( c( 2, "3", 4)))
te$test( "character"==mode( c( 2, "3", 4, TRUE, NULL)))
te$test( "character"==typeof( c( 2, "3", 4, TRUE, NULL)))
te$test( "character"==class( c( 2, "3", 4, TRUE, NULL)))

te$test( "list" == mode( list()))
te$test( "list" == typeof( list()))
te$test( "list" == mode( list( 1L, 1.1, NULL, TRUE, 'abc')))
te$test( "list" == typeof( list( 1L, 1.1, NULL, TRUE, 'abc')))
te$test( "list" == class( list( 1L, 1.1, NULL, TRUE, 'abc')))

te$test( "call"==mode( y ~ a + b))
te$test( "language"==typeof( y ~ a + b))
te$test( "formula"==class( y ~ a + b))

te$test( "2" == c( 2, "3", 4, NULL)[ 1])
te$test( 2 == c( 2, "3", 4, NULL)[ 1])
#te$test( c( 2, "3", 4, NULL) == c( 2, "3", 4, NULL)[ 0]) # Argument hat Laenge 0
#te$test( NULL != NULL) # Argument hat Laenge 0
te$test( "2" == c( 2, "3", 4, NULL)[ 1])
te$test( "TRUE" == c( 2, "3", 4, TRUE)[ 4])
te$test( TRUE == c( 2, "3", 4, TRUE)[ 4])


te$test( TRUE == "TRUE")
te$test( TRUE == 1)
te$test( "TRUE" != 1)

te$test( 2 == "2")

te$test( is.null( NULL))

te$test( 1 == as.numeric( TRUE))
te$test( 0 == as.numeric( FALSE))

te$test( 1 == c( 2, 3, 4, TRUE)[ 4])

te$test( all( logical( 3) == c( FALSE, FALSE, FALSE)))

te$test( all( numeric( 4) == c( 0, 0, 0, 0)))

te$test( all( character( 5) == c( '', '', '', '', '')))


te$checkComplainAndAdjustExpected( 61)

print( te$result())

File: ocaml-4.01.0/topic_types.ml

let 
 checkComplainAndAdjustExpected= Testenv.TestEnv.checkComplainAndAdjustExpected and
 test= Testenv.TestEnv.test and
 result= Testenv.TestEnv.result;;

checkComplainAndAdjustExpected 0;;

type ab= A | B;;
let x= B;;

match x with 
| unbound -> test true
;;

match x with 
| unbound -> test (unbound == x)
;;

match x with 
| unbound -> test (unbound == B)
;;

match x with 
| (_ : ab) -> test true
;;

match x with 
| (B : ab) -> test true
| _ -> test false
;;

checkComplainAndAdjustExpected 5;;

type bc= B | C;;
let y= B;;

(* test (x=y);; *) (* leads to error *)
(* no construction of B from type ab possible *)

match x with 
| _ -> test true
;;

match x with 
| _ -> test true
;;

match x with 
| (_) -> test true
;;

match x with 
| ab -> test true
;;

match y with 
| bc -> test true
;;

checkComplainAndAdjustExpected 5;;

type i= { i : int };;

let f ({i=x}:i)= x;;
test (1==f {i=1});;

let i= {i= 1};;

match i with
| unbound -> test (unbound = {i= 1})
;;

match i with
| {i=unbound} -> test (unbound == 1)
;;

match i with
| {i= 2} -> test false
| {i= 1} -> test true
| _ -> test false
;;

checkComplainAndAdjustExpected 4;;

type k= I of int;;

let f (I(x):k) = x;;
test (1==f (I(1)));;

let i= I( 1);;

match i with
| unbound -> test (unbound = I( 1))
;;

match i with
| I(unbound) -> test (unbound == 1)
;;

match i with
| I(_) -> test true
;;

match i with
| I( 2) -> test false
| I( 1) -> test true
| _ -> test false
;;

checkComplainAndAdjustExpected 5;;


type is= { i : int; s : string};;
type l= I of int * is | K of int*int*string;;

let f (I(x,{i=y; s=z})|K (x, y, z) :l)= (x, y, z);;
test ((1, 2, "3")=f (I(1,{i=2;s="3"})));;
test ((1, 2, "3")=f (K(1,2,"3")));;

let f = function
| I(x,{i=y; s=z}) -> (x, y, z)
| K (x, y, z) -> (x, y, z)
;;

test ((1, 2, "3")=f (I(1,{i=2;s="3"})));;
test ((1, 2, "3")=f (K(1,2,"3")));;

(*type l= I of int * is | K;;*)

let l= I(2,{i=3; s="4"});;

match l with
| unbound -> test (unbound = I(2,{i=3; s="4"}))
;;


match l with
| I(u1,u2) -> test ( u1 == 2 && u2 = {i=3; s="4"})
| K(_) -> test false
;;

match l with
| I(u1,{i=u2; s=u3}) -> test (u1 == 2 && u2 == 3 && u3 = "4")
| K(_) -> test false
;;

checkComplainAndAdjustExpected 7;;

Printf.printf "%s\n" (result ());;



